File: packages/python-sdk/prompttrace/storage.py
import os
import json
from typing import Dict

class StorageEngine:
    def __init__(self):
        self.dir_path = self._resolve_project_root()
        # Changed to JSONL for append-only concurrent safety
        self.file_path = os.path.join(self.dir_path, 'traces.jsonl')
        self.memory_hash_map: Dict[str, int] = {}
        self._initialize()

    def _resolve_project_root(self) -> str:
        current_dir = os.getcwd()
        # Common project indicators enforcing a boundary state mirroring the TS 'package.json' fallback
        indicators = ['pyproject.toml', 'requirements.txt', 'setup.py', '.git', 'package.json']
        
        while True:
            for ind in indicators:
                if os.path.exists(os.path.join(current_dir, ind)):
                    return os.path.join(current_dir, '.prompttrace')
            
            parent = os.path.dirname(current_dir)
            if parent == current_dir:
                # Ultimate fallback gracefully isolating to root execution
                return os.path.join(os.getcwd(), '.prompttrace')
            
            current_dir = parent

    def _initialize(self):
        try:
            if not os.path.exists(self.dir_path):
                os.makedirs(self.dir_path, exist_ok=True)
            if not os.path.exists(self.file_path):
                with open(self.file_path, 'w') as f:
                    f.write('')
            else:
                # O(1) scale: no longer parsing the full file to populate memory cache.
                pass
        except Exception as e:
            print(f"[Prompttrace] Failed to initialize storage: {e}")

    def save_trace(self, trace_dict: dict):
        try:
            # Append-only operation ensuring cross-language thread safety
            line = json.dumps(trace_dict) + '\n'
            with open(self.file_path, 'a') as f:
                f.write(line)
                f.flush()
        except Exception as e:
            print(f"[Prompttrace] Failed to save trace: {e}")

    def register_and_get_hit_count(self, prompt_hash: str) -> int:
        hits = self.memory_hash_map.get(prompt_hash, 0)
        self.memory_hash_map[prompt_hash] = hits + 1
        return hits

File: packages/sdk/src/storage.ts
import fs from 'fs';
import path from 'path';
import { TraceRecord } from './types';

export class StorageEngine {
  private dirPath: string;
  private filePath: string;
  private memoryHashMap: Map<string, number>;

  constructor() {
    const root = this.resolveRootPath();

    this.dirPath = path.join(root, '.prompttrace');
    // Changed to JSONL for append-only concurrent safety
    this.filePath = path.join(this.dirPath, 'traces.jsonl');
    this.memoryHashMap = new Map();

    this.initialize();
  }

  /**
   * Walk up directories until we find project root (package.json)
   */
  private resolveRootPath(): string {
    let currentDir = process.cwd();

    while (true) {
      const possible = path.join(currentDir, 'package.json');

      if (fs.existsSync(possible)) {
        return currentDir;
      }

      const parent = path.dirname(currentDir);

      // reached filesystem root → fallback
      if (parent === currentDir) {
        console.warn('[Prompttrace] Could not find project root, using cwd');
        return process.cwd();
      }

      currentDir = parent;
    }
  }

  private initialize() {
    try {
      if (!fs.existsSync(this.dirPath)) {
        fs.mkdirSync(this.dirPath, { recursive: true });
      }

      if (!fs.existsSync(this.filePath)) {
        // Create empty jsonl file
        fs.writeFileSync(this.filePath, '');
      } else {
        // Note: For O(1) performance and scaling, we NO LONGER parse the entire file into memory 
        // to populate the memoryHashMap. It now operates as a runtime-only fast cache.
      }
    } catch (err) {
      console.error('[Prompttrace] Failed to initialize storage:', err);
    }
  }

  public saveTrace(trace: TraceRecord) {
    try {
      // Append-only JSONL format guarantees no corruption during concurrent TS/Python writes
      const line = JSON.stringify(trace) + '\n';
      fs.appendFileSync(this.filePath, line);

      // 🔥 Important: helps chokidar detect changes more reliably
      fs.utimesSync(this.filePath, new Date(), new Date());
    } catch (err) {
      console.error('[Prompttrace] Failed to save trace:', err);
    }
  }

  public registerAndGetHitCount(hash: string): number {
    const hits = this.memoryHashMap.get(hash) || 0;
    this.memoryHashMap.set(hash, hits + 1);
    return hits;
  }
}